From 023cbf23dae5e71ccebdf7c37632c36b423907e9 Mon Sep 17 00:00:00 2001
From: Zhantao Tang <zhantao.tang@windriver.com>
Date: Thu, 8 Apr 2021 10:28:00 +0800
Subject: [PATCH] secboot: add key store status check point after sys_image
 check

The key stores may need to reformat even if the sys_image
exists. So adding one more check point after sys_image check,
this will guarantee the normal work of HSE firmware, otherwise,
the hse demo in linux may fail due to the unformatted status of
hse key stores.

Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 arch/arm/cpu/armv8/s32/s32-gen1/hse_adv_secboot.c | 11 +++++++++++
 include/hse/hse_abi.h                             |  1 +
 2 files changed, 12 insertions(+)

diff --git a/arch/arm/cpu/armv8/s32/s32-gen1/hse_adv_secboot.c b/arch/arm/cpu/armv8/s32/s32-gen1/hse_adv_secboot.c
index ff423f045c..6bac060af2 100644
--- a/arch/arm/cpu/armv8/s32/s32-gen1/hse_adv_secboot.c
+++ b/arch/arm/cpu/armv8/s32/s32-gen1/hse_adv_secboot.c
@@ -563,6 +563,17 @@ static int do_hse_keystore_format(cmd_tbl_t *cmdtp, int flag,
 	/* check if sys_img already exists */
 	if (priv->ivt.sys_img) {
 		printf("CHECK: SYS_IMG already exists\n");
+
+		if (!(hse_status_ret & HSE_STATUS_INSTALL_OK)) {
+			/* if HSE_STATUS_INSTALL_OK bit not set, then need to reformat key stores */
+			printf("CHECK: key stores need to reformat!\n");
+			ret = hse_format_key_store(priv, &hse_recv);
+			if (ret)
+				goto ret_fail;
+
+			printf("CHECK: key stores formatted success!\n");
+		}
+
 		ret = CMD_RET_SUCCESS;
 		goto ret_fail;
 	}
diff --git a/include/hse/hse_abi.h b/include/hse/hse_abi.h
index d29f8168d3..29f60ca4a2 100644
--- a/include/hse/hse_abi.h
+++ b/include/hse/hse_abi.h
@@ -39,6 +39,7 @@
 #define HSE_CHANNEL_GENERAL 1u
 
 #define HSE_STATUS_INIT_OK  BIT(8)
+#define HSE_STATUS_INSTALL_OK  BIT(9)
 #define HSE_IVT_BOOTSEQ_BIT BIT(3)
 
 #define HSE_SRV_ID_SET_ATTR                 0x00000001ul
-- 
2.17.1

